lib/rust/ensogl/component/grid-view/src/visible_area.rs

```diff
@@ -38,18 +38,39 @@ pub fn visible_columns(
     col_count: usize,
     column_widths: &ColumnWidths,
 ) -> Range<Col> {
-    let column_left = |idx: usize| idx as f32 * entry_size.x + column_widths.pos_offset(idx);
+    let entry_width = entry_size.x;
+    let column_left = |idx: usize| idx as f32 * entry_width + column_widths.pos_offset(idx);
     let column_right = |idx: usize| column_left(idx + 1);
     let index = |(idx, _): (usize, _)| idx;
-
-    let visible = |(_, x): &(_, f32)| v.left < *x;
-    let mut right_borders = (0..col_count).map(|idx| (idx, column_right(idx)));
-    let first_visible = right_borders.find(visible).map(index).unwrap_or(col_count);
-
+    let clamp = |idx: isize| idx.clamp(0, col_count as isize) as Col;
+
+    let left_guess = clamp((v.left / entry_width).floor() as isize);
+    let first_visible = {
+        let pos_offset = column_widths.pos_offset(left_guess);
+        let column_left_border = left_guess as f32 * entry_width + pos_offset;
+        let column_width = entry_width + column_widths.width_diff(left_guess);
+        let column_right_border = column_left_border + column_width;
+        let is_visible = column_right_border > v.left && column_left_border < v.right;
+        let not_shifted = pos_offset == 0.0;
+
+        if not_shifted && is_visible {
+            left_guess
```

Comment:
Actually, the left_guess could be shifted and still on the edge. Instead of `is_visible` checking, I would check if `column_left_border <= v.left && column_right_border >= v.left`.

Or, if you prefer simple code, we don't have to check visibility? I deem If it's not shifted, it's always visible.

Reply:
It can be shifted to the right and be still visible, but it won't be the **first** visible column. A proposed `column_left_border <= v.left && column_right_border >= v.left` doesn't work for the same reason.

> I deem If it's not shifted, it's always visible.

This is also not correct, unfortunately. If a column is shrunk, its position is unchanged (it has pos_offset of zero), but it still can be behind the left viewport border in some cases.